cmd/otk-resolve-ostree-commit/main_test.go

```diff
@@ -0,0 +1,104 @@
+package main_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	resolver "github.com/osbuild/images/cmd/otk-resolve-ostree-commit"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Create a test server that responds with the commit ID that corresponds to
+// the ref.
+func createTestServer(refIDs map[string]string) *httptest.Server {
+	handler := http.NewServeMux()
+	handler.HandleFunc("/refs/heads/", func(w http.ResponseWriter, r *http.Request) {
+		reqRef := strings.TrimPrefix(r.URL.Path, "/refs/heads/")
+		id, ok := refIDs[reqRef]
+		if !ok {
+			http.NotFound(w, r)
+			return
+		}
+		fmt.Fprint(w, id)
+	})
+
+	return httptest.NewServer(handler)
+}
+
+func TestResolver(t *testing.T) {
+	commitMap := map[string]string{
```

Comment:
We could get away with a simpler test here I think. It's fine as is given that it's all written etc but I think at this level here all we care about and need to test is that we can read the inputs, call `ostree.Resolve()` with the right inputs and then return the expected outputs. So just doing something like:
```go
diff --git a/cmd/otk-resolve-ostree-commit/export_test.go b/cmd/otk-resolve-ostree-commit/export_test.go
index 77c5f1a93..91ab63416 100644
--- a/cmd/otk-resolve-ostree-commit/export_test.go
+++ b/cmd/otk-resolve-ostree-commit/export_test.go
@@ -3,3 +3,12 @@ package main
 var (
        Run = run
 )
+
+func MockOstreeResolve(new func(...)) (restore func()) {
+       saved := ostreeResolve
+       ostreeResolve = new
+       return func() {
+               ostreeResolved = saved
+       }
+}
+
diff --git a/cmd/otk-resolve-ostree-commit/main.go b/cmd/otk-resolve-ostree-commit/main.go
index 3857dfd37..a7ea2dd5e 100644
--- a/cmd/otk-resolve-ostree-commit/main.go
+++ b/cmd/otk-resolve-ostree-commit/main.go
@@ -44,6 +44,8 @@ type Output struct {
        Checksum string `json:"checksum"`
 }
 
+var ostreeResolve = ostree.Resolve
+
 func run(r io.Reader, w io.Writer) error {
        var inputTree Tree
        if err := json.NewDecoder(r).Decode(&inputTree); err != nil {
@@ -51,7 +53,7 @@ func run(r io.Reader, w io.Writer) error {
        }
 
        sourceSpec := ostree.SourceSpec(inputTree.Tree)
-       commitSpec, err := ostree.Resolve(sourceSpec)
+       commitSpec, err := ostreeResolve(sourceSpec)
        if err != nil {
                return fmt.Errorf("failed to resolve ostree commit: %w", err)
        }

```
(the upside is that we don't need to care about implementation details of ostree.Resolve() at this level and we just trust that it's well tested to reach out to the net etc and if implementation details change we only need to tweak the tests in ostree and here we are blisfully ignorant (which again of course here is not very important as this behavior/test will most likely never change))

Reply:
There's some logic in the resolver that I'd rather not mock.  The resolver is tested elsewhere, yes, but I'd also like for the external to test its own expectations and assumptions.  The interaction between the arguments for example is something that might change and it'd be good to have those tested in the external.  Mocking the function would make us miss those changes.  On the other hand, mocking the server itself is perfect because it's so simple, it's just a single GET with a very easy to fake response.  I think given how cheap it is to catch regressions and changes in the `ostree.Resolve()` call for the external, mocking the server and not the function is a lot more beneficial.

Reply:
To summarise: With the mock of the server we are in fact testing `ostree.Resolve()`, but only as a means of making sure that the `otk-resolve-ostree-commit` tool is preparing the right inputs.  The validity of a given input can change without notice if we mock `ostree.Resolve()`.

Reply:
I'm may misunderstand the argument (my apologies if I did) - once we release this tool the inputs for this tool become an API that will be used in many otk manifests. That means that any inputs for `ostree.Resolve()` are now also an API, we cannot restrict what was valid before anymore (also afaik osbuild composer uses `ostree.Resolve()` too?).  Again, totally fine to leave as is (and stop this thread) I'm really just trying to understand (and sorry for the extra work my questions cause).

Reply:
The arguments to `ostree.Resolve()` become API indirectly, but we might change that function's behaviour.

Here's a concrete example that we might actually do at some point:

Let's say we decide that the `ref` argument to `ostree.Resolve()` **must** be a valid ref (similar to a branch name) and cannot be an ostree commit ID (sha256 checksum).  After this change, passing a checksum to the resolver will result in an error.  This makes it the job of the caller to check what it's passing through to the resolver, basically moving this block outside the function: https://github.com/achilleas-k/images/blob/31d94be4ec71566036c397b43deeffb349c71ba7/pkg/ostree/ostree.go#L238-L242

If that happens and we want to keep the otk external's API consistent, we would run the checksum check in the `run()` function of the otk external before calling `ostree.Resolve()`.  The external's API and behaviour don't change even though the resolver function's did.

Without calling the real `ostree.Resolve()` in tests, no test would fail from this change.